File: src/main/cpp/commands/command_DriveTeleop.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/command_DriveTeleop.h"
#include "frc/smartdashboard/SmartDashboard.h"
command_DriveTeleop::command_DriveTeleop(subsystem_DriveTrain* DriveTrain, subsystem_PoseTracker* PoseTracker,
                                          std::function<double()> xSpeed,
                                          std::function<double()> ySpeed,
                                          std::function<double()> zRotation,
                                          std::function<double()> IsOrientFront,
                                          std::function<double()> IsOrientBack,
                                          std::function<bool()> FieldRelative,
                                          std::function<bool()> OpenLoop): m_DriveTrain{DriveTrain}, m_PoseTracker{PoseTracker},
                                                      m_xSpeed{xSpeed},
                                                      m_ySpeed{ySpeed},
                                                      m_zRotation{zRotation},
                                                      m_IsOrientFront{IsOrientFront},
                                                      m_IsOrientBack{IsOrientBack},
                                                      m_FieldRelative{FieldRelative},
                                                      m_OpenLoop{OpenLoop}
{
  // Use addRequirements() here to declare subsystem dependencies.
  AddRequirements({m_DriveTrain});
  AddRequirements({m_PoseTracker});
}

// Called when the command is initially scheduled.
void command_DriveTeleop::Initialize() {}

// Called repeatedly when this Command is scheduled to run
void command_DriveTeleop::Execute() {

    // m_DriveTrain->ImplementVisionPose(m_PoseTracker->GetEstimatedGlobalPose());

    // std::pair<frc::Pose2d, units::second_t> Pose = m_PoseTracker->GetEstimatedGlobalPose().;
    // frc::SmartDashboard::PutNumber("VisionX")

  

  m_DriveTrain -> SetAutoOrient(m_IsOrientFront(), m_IsOrientBack(), m_zRotation());

  double TransformedXSpeed = 0.0;
  double TransformedYSpeed = 0.0;
  double TransformedZRot = 0.0;
  if(fabs(m_xSpeed()) > ControllerConstants::Deadband){
    TransformedXSpeed = m_DriveTrain->SetThrottle(m_xSpeed());
  }

  if(fabs(m_ySpeed()) > ControllerConstants::Deadband){
    TransformedYSpeed = m_DriveTrain->SetThrottle(m_ySpeed());
  }

  if(fabs(m_zRotation()) > ControllerConstants::Deadband){
    TransformedZRot = m_DriveTrain->SetThrottle(m_zRotation());
  }


  


  
  
  // m_DriveTrain -> SwerveDrive( m_DriveTrain -> SetThrottle( frc::ApplyDeadband(m_xSpeed(), ControllerConstants::Deadband) )* SwerveConstants::MaxSpeed,
  //                               m_DriveTrain -> SetThrottle( frc::ApplyDeadband(m_ySpeed(), ControllerConstants::Deadband) ) * SwerveConstants::MaxSpeed,
  //                               m_DriveTrain -> SetThrottle( frc::ApplyDeadband(m_zRotation(), ControllerConstants::Deadband) ) * SwerveConstants::MaxAngularVelocity,
  //                               m_FieldRelative(),
  //                               m_OpenLoop());
  m_DriveTrain -> SwerveDrive( TransformedXSpeed *  SwerveConstants::MaxSpeed,
                                TransformedYSpeed * SwerveConstants::MaxSpeed,
                                TransformedZRot * SwerveConstants::MaxAngularVelocity,
                                m_FieldRelative(),
                                m_OpenLoop());



  
  

}

// Called once the command ends or is interrupted.
void command_DriveTeleop::End(bool interrupted) {}

// Returns true when the command should end.
bool command_DriveTeleop::IsFinished() {
  return false;
}

File: src/main/java/frc/robot/commands/command_DriveTeleop.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import java.util.function.BooleanSupplier;
import java.util.function.DoubleSupplier;

import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.robot.Constants.ControllerConstants;
import frc.robot.Constants.SwerveConstants;
import frc.robot.subsystems.subsystem_DriveTrain;

public class command_DriveTeleop extends CommandBase {
  /** Creates a new command_DriveTeleop. */
  private subsystem_DriveTrain s_DriveTrain;
  private DoubleSupplier m_xSpeed;
  private DoubleSupplier m_ySpeed;
  private DoubleSupplier m_zRot;
  private BooleanSupplier m_IsOrientFront;
  private BooleanSupplier m_IsOrientBack;
  private BooleanSupplier m_FieldRelative;
  private BooleanSupplier m_OpenLoop;

  public command_DriveTeleop(subsystem_DriveTrain driveTrain,
                            DoubleSupplier xSpeed,
                            DoubleSupplier ySpeed,
                            DoubleSupplier zRot,
                            BooleanSupplier isOrientFront,
                            BooleanSupplier isOrientBack,
                            BooleanSupplier fieldRelative,
                            BooleanSupplier openLoop) {
    // Use addRequirements() here to declare subsystem dependencies.
    s_DriveTrain = driveTrain;
    m_xSpeed = xSpeed;
    m_ySpeed = ySpeed;
    m_zRot = zRot;
    m_IsOrientFront = isOrientFront;
    m_IsOrientBack = isOrientBack;
    m_FieldRelative = fieldRelative;
    m_OpenLoop = openLoop;
    addRequirements(s_DriveTrain);
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {}

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    s_DriveTrain.setAutoOrient(m_IsOrientFront.getAsBoolean(), m_IsOrientBack.getAsBoolean(), m_zRot.getAsDouble());
    
    double transformedXSpeed = 0.0;
    double transformedYSpeed = 0.0;
    double transformedZRot = 0.0;

    if(Math.abs(m_xSpeed.getAsDouble()) > ControllerConstants.deadband){
      transformedXSpeed = s_DriveTrain.setThrottle(m_xSpeed.getAsDouble());
    }
    if(Math.abs(m_ySpeed.getAsDouble()) > ControllerConstants.deadband){
      transformedYSpeed = s_DriveTrain.setThrottle(m_ySpeed.getAsDouble());
    }
    if(Math.abs(m_zRot.getAsDouble()) > ControllerConstants.deadband){
      transformedZRot = s_DriveTrain.setThrottle(m_zRot.getAsDouble());
    }

    s_DriveTrain.swerveDrive(transformedXSpeed * SwerveConstants.maxSpeed,
                            transformedYSpeed * SwerveConstants.maxSpeed,
                            transformedZRot * SwerveConstants.maxAngularVelocity,
                            m_FieldRelative.getAsBoolean(),
                            m_OpenLoop.getAsBoolean());
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {}

  // Returns true when the command should end.
  @Override
  public boolean isFinished() {
    return false;
  }
}
